framework/src/iotbus/iotbus_gpio.c

```diff
@@ -400,88 +412,142 @@ int iotbus_gpio_unregister_cb(iotbus_gpio_context_h dev)
+		return IOTBUS_ERROR_INVALID_PARAMETER;
 	}
 
-	ret = ioctl(fd, IOTBUS_INTR_REGISTER, (unsigned long)&info);
-	if (ret < 0) {
-		ibdbg("Fail to register %d\n", int_type);
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+	switch (int_type) {
+	case IOTBUS_GPIO_EDGE_BOTH:
+		return IOTBUS_ERROR_NOT_SUPPORTED;
+	case IOTBUS_GPIO_EDGE_RISING:
+		notify.gn_rising  = true;
+		notify.gn_falling = false;
+		notify.gn_signo = SIG_IOTBUS_GPIO_RISING;
+		break;
+	case IOTBUS_GPIO_EDGE_FALLING:
+		notify.gn_rising  = false;
+		notify.gn_falling = true;
+		notify.gn_signo = SIG_IOTBUS_GPIO_FALLING;
+		break;
+	default:
+		return IOTBUS_ERROR_INVALID_PARAMETER;
 	}
-	ret = ioctl(handle->fd, GPIOIOC_SET_INTERRUPT, edge);
-	if (ret < 0) {
-		ibdbg("Fail to set %d\n", int_type);
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+
+	struct intr_attr val = { priority, IOTBUS_GPIO, getpid() };
+
+	handle->cb = cb;
+	handle->pid = create_intr_pthread(gpio_intr_thread, (void *)dev, &val);
+	if (handle->pid <= 0) {
+		handle->cb = NULL;
+		handle->pid = 0;
+		return IOTBUS_ERROR_UNKNOWN;
 	}
+
+	notify.pid = handle->pid;
+	notify.handle = (void *)dev;
 	
-	close(fd);
+	ret = ioctl(handle->fd, GPIOIOC_SET_INTERRUPT, (unsigned long)&notify);
+	if (ret != 0) {
+		ibdbg("ioctl failed \n");
+		handle->cb = NULL;
+		if (pthread_cancel(handle->pid) < 0) {
+			ibdbg("pthread cancel failed\n");
+			return IOTBUS_ERROR_UNKNOWN;
+		}
+		handle->pid = 0;
+		return ret;
+	}
 	return IOTBUS_ERROR_NONE;
 }
 
-int iotbus_gpio_unset_interrupt(iotbus_gpio_context_h dev, iotbus_int_type_e int_type)
+int iotbus_gpio_unset_interrupt(iotbus_gpio_context_h dev, iotbus_gpio_edge_e int_type)
 {
-	int fd, ret, edge;
+	/* 
+	1. Diable Interrupt
+	2. Kill Thread
+	3. Clear Callback
+	*/
 	struct _iotbus_gpio_s *handle;
-	struct iotbus_int_info_s info = { 0, };
+	FAR struct gpio_notify_s notify;
+	int ret;
+
+	if (!dev || !dev->handle) {
+		return IOTBUS_ERROR_INVALID_PARAMETER;
+	}
 
 	handle = (struct _iotbus_gpio_s *)dev->handle;
 
-	info.handle = dev;
-	info.pin_type = IOTBUS_GPIO;
-	info.int_type = int_type;
-	
-	fd = open(IOTBUS_SIGPATH, O_NONBLOCK);
-	if (fd < 0) {
-		ibdbg("Open iotsig fail[%d, %d]\n", fd, get_errno());
-		return IOTBUS_ERROR_DEVICE_FAIL;
-	}
+	// Disable Interrupt
+	notify.gn_rising  = false;
+	notify.gn_falling = false;
+	notify.gn_signo = 0;
+	notify.pid = 0;
+	notify.handle = NULL;
 
-	ret = ioctl(handle->fd, GPIOIOC_SET_INTERRUPT, GPIO_EDGE_NONE);
-	if (ret < 0) {
-		ibdbg("Fail to unset intterupt\n");
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+	ret = ioctl(handle->fd, GPIOIOC_SET_INTERRUPT, (unsigned long)&notify);
+	if (ret != 0) {
+		ibdbg("ioctl failed \n");
+		return ret;
 	}
-	ret = ioctl(fd, IOTBUS_INTR_UNREGISTER, (unsigned long)&info);
-	if (ret < 0) {
-		ibdbg("Fail to unregister %d\n", int_type);
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+
+	// Kill Thread
+	if (handle->pid > 0) {
+		ret = pthread_cancel(handle->pid);
```

Comment:
I guess timing issue can be occured between disable notification from driver & disable pthread.
I think calling pthread_cancel first is better, but I am not sure. How do you think??